packages/ra-ui-materialui/src/preferences/Configurable.tsx

```diff
@@ -0,0 +1,154 @@
+import * as React from 'react';
+import { useRef, useEffect, cloneElement, ReactElement } from 'react';
+import { usePreferencesEditor, PreferenceKeyContextProvider } from 'ra-core';
+import { alpha, Badge } from '@mui/material';
+import { styled, SxProps } from '@mui/material/styles';
+import SettingsIcon from '@mui/icons-material/Settings';
+import clsx from 'clsx';
+
+/**
+ * Wrap any component with this component to make it configurable
+ *
+ * When the edit mode is enabled, users will see a button to edit the component;
+ * when clicked, the inspector will show the editor element.
+ *
+ * Creates a context for the preference key, so that both the child component
+ * and the editor can access it using usePreferenceKey();
+ *
+ * @example
+ * const ConfigurableTextBlock = ({ preferenceKey = "TextBlock", ...props }) => (
+ *     <Configurable editor={<TextBlockInspector />} preferenceKey={preferenceKey}>
+ *         <TextBlock {...props} />
+ *     </Configurable>
+ * );
+ */
+export const Configurable = (props: ConfigurableProps) => {
+    const {
+        children,
+        editor,
+        preferenceKey,
+        openButtonLabel = 'ra.configurable.customize',
+        sx,
+    } = props;
+
+    const prefixedPreferenceKey = `preferences.${preferenceKey}`;
+    const preferencesEditorContext = usePreferencesEditor();
+    const hasPreferencesEditorContext = !!preferencesEditorContext;
+
+    const {
+        isEnabled,
+        setEditor,
+        preferenceKey: currentPreferenceKey,
+        setPreferenceKey,
+    } = preferencesEditorContext || {};
+
+    const isEditorOpen = prefixedPreferenceKey === currentPreferenceKey;
+    const editorOpenRef = useRef(isEditorOpen);
+
+    useEffect(() => {
+        editorOpenRef.current = isEditorOpen;
+    }, [isEditorOpen]);
+
+    // on unmount, if selected, remove the editor
+    useEffect(() => {
+        return () => {
+            if (!editorOpenRef.current) return;
+            setPreferenceKey && setPreferenceKey(null);
+            setEditor && setEditor(null);
+        };
+    }, [setEditor, setPreferenceKey]);
+
+    if (!hasPreferencesEditorContext) {
+        return children;
+    }
```

Comment:
Should be moved line 37 (early return)

Reply:
No can't do: because of the rules of hooks, hooks can't be executed conditionally.